### PR TITLE
[fix #2985] Show sent transaction immediatly in history

### LINKED ISSUE #2985: Transaction is not immediately visible in Transactions History
### Description

*Type*: Bug

*Summary*: When a user sends ETH and then goes to Transactions History the transaction is not visible there. User needs to wait until the transaction has at least couple of confirmations to see it in history. Not seeing the transaction immediately may lead him to think that something went wrong.  

#### Expected behavior
Transaction that was sent is immediately visible in Transactions History. Or, at least there is an indicator that transaction is being processed.

#### Actual behavior
Transaction becomes visible after couple of confirmations.

### Reproduction
[comment]: # (Describe how we can replicate the bug step by step.)

- Open Status
- Go to Wallet 
- Send a transaction and go to Transactions History

### Additional Information
* Status version: 0.9.12 (TestFlight)
* Operating System: iOS
* Tested on: iPhone 6, iOS 11.2.1

### PR DESCRIPTION
**Pull request by yenda**
fixes #2985 

### Steps to test:
- Open Status
- Send transaction and go immediatly in transaction history
- Verify that transaction is there, check transaction details
- Wait a bit and refresh transaction history
- Verify that transaction details have updated with new information available

Previous behavior:
Transaction was not visible at all in transaction history until it could be seen on etherscan

status: ready

### CODE REVIEW

**Review by janherich — APPROVED** [review]

File: src/status_im/ui/screens/wallet/send/events.cljs [hunk 1]
```diff
@@ -265,4 +277,4 @@
 (handlers/register-handler-fx
   :wallet.send/set-gas-price
   (fn [{:keys [db]} [_ gas-price]]
-    {:db (assoc-in db [:wallet :send-transaction :gas-price] (money/bignumber gas-price))}))
\ No newline at end of file
+    {:db (assoc-in db [:wallet :send-transaction :gas-price] (money/bignumber gas-price))}))
```
Comment (jeluard) [inline src/status_im/ui/screens/wallet/send/events.cljs:281]: Can you exclude those changes from your commits? Looks like it's back and forth and sometimes causes git to notice changes on my env.

File: src/status_im/ui/screens/wallet/send/events.cljs [hunk 2]
```diff
@@ -162,19 +163,30 @@
...
-        (merge
-          {:db (-> db'
-                   (update-in [:wallet :transactions-unsigned] dissoc id)
-                   (update-in [:wallet :send-transaction] merge clear-send-properties))}
-          (if modal?
-            {:dispatch-n [[:navigate-back]
-                          [:navigate-to-modal :wallet-transaction-sent-modal]]}
-            {:dispatch [:navigate-to :wallet-transaction-sent]}))))))
+        (let [hash        (:hash response)
+              transaction (-> (get-in db [:wallet :transactions-unsigned id])
```
Comment (jeluard) [inline src/status_im/ui/screens/wallet/send/events.cljs]: Could we extract this as a function to improve lisibility?

**Review by jeluard — CHANGES_REQUESTED** [review]

File: src/status_im/ui/screens/wallet/send/events.cljs (hunk 1 above)
Comment (yenda) [reply src/status_im/ui/screens/wallet/send/events.cljs:281]: having newline character at the end of files is the correct convention, if your editor removes it configure it. That is the reason why git shows it like this when it is missing.

Comment (jeluard) [reply src/status_im/ui/screens/wallet/send/events.cljs:281]: The correct convention is the one we agree on :)
I have a similar issue with `package-lock.json` and I don't think I touch this file at all.

**Comment by yenda** [comment]
@jeluard @janherich I simplified the PR

**Review by jeluard — APPROVED** [review]

File: src/status_im/ui/screens/wallet/events.cljs [hunk 3]
```diff
@@ -124,7 +124,7 @@
   :update-transactions-success
   (fn [db [_ transactions]]
     (-> db
-        (assoc-in [:wallet :transactions] transactions)
+        (update-in [:wallet :transactions] #(merge % transactions))
```
Comment (dmitryn) [inline src/status_im/ui/screens/wallet/events.cljs]: Could be just `(update-in [:wallet :transactions] merge transactions)`

File: src/status_im/ui/screens/wallet/send/events.cljs [hunk 4]
```diff
@@ -160,21 +161,34 @@
            ::show-transaction-error error_message}
           {:db (update-in db [:wallet :transactions-unsigned] dissoc id)})))))
 
+(defn prepare-unconfirmed-transaction [db now hash id]
+  (-> (get-in db [:wallet :transactions-unsigned id])
```
Comment (dmitryn) [inline src/status_im/ui/screens/wallet/send/events.cljs]: I would prefer
```
(-> db
    (get-in [:wallet :transactions-unsigned id])
    (assoc ...
```

File: src/status_im/ui/screens/wallet/send/events.cljs [hunk 5]
```diff
@@ -160,21 +161,34 @@
...
         {:db db'}
-        (merge
-          {:db (-> db'
-                   (update-in [:wallet :transactions-unsigned] dissoc id)
-                   (update-in [:wallet :send-transaction] merge clear-send-properties))}
-          (if modal?
-            {:dispatch-n [[:navigate-back]
-                          [:navigate-to-modal :wallet-transaction-sent-modal]]}
-            {:dispatch [:navigate-to :wallet-transaction-sent]}))))))
+        (let [hash (:hash response)]
```
Comment (dmitryn) [inline src/status_im/ui/screens/wallet/send/events.cljs]: Did we already capture `hash` in `let` on line 178 ?

**Review by goranjovic — APPROVED** [review]

**Review by dmitryn — APPROVED** [review]

File: src/status_im/ui/screens/wallet/send/events.cljs (hunk 4 above)
Comment (yenda) [reply src/status_im/ui/screens/wallet/send/events.cljs]: but it is better to have only one kind of data in a threading macro, so the first line gets a transaction from the db and next lines are changing this transaction. If you separate the first line then the first line you have a db, then a transaction, then you change this transaction

File: src/status_im/ui/screens/wallet/events.cljs (hunk 3 above)
Comment (yenda) [reply src/status_im/ui/screens/wallet/events.cljs]: If you do this then the transactions that are not confirmed will never be updated. `%` is the current transaction history (including the one that was derived from the unsigned transaction and is not yet confirmed in the chain) so you want its values to be overriden by `transactions` which is data from etherscan when they exist

File: src/status_im/ui/screens/wallet/send/events.cljs (hunk 5 above)
Comment (yenda) [reply src/status_im/ui/screens/wallet/send/events.cljs]: thanks good catch !

File: src/status_im/ui/screens/wallet/events.cljs (hunk 3 above)
Comment (dmitryn) [reply src/status_im/ui/screens/wallet/events.cljs]: @yenda i'm just saying that 
`(update-in [:wallet :transactions] merge transactions)` is equal to
`(update-in [:wallet :transactions] #(merge % transactions))`
but shorter, no need to create anonymous fn

File: src/status_im/ui/screens/wallet/send/events.cljs (hunk 4 above)
Comment (dmitryn) [reply src/status_im/ui/screens/wallet/send/events.cljs]: Didn't get your point, but it seems like style differences for me. Usually i start threading macro with data structure i'm working on, then apply transformation/extraction step-by-step. In your case, first line of threading macro is already an operation. Consider this example from 
https://github.com/bbatsov/clojure-style-guide
```
(-> [1 2 3]
    reverse
    (conj 4)
    prn)
```
or
```
(-> (reverse [1 2 3])
    (conj 4)
    prn)
```

File: src/status_im/ui/screens/wallet/events.cljs (hunk 3 above)
Comment (yenda) [reply src/status_im/ui/screens/wallet/events.cljs]: yes you are correct, fixed, thanks

File: src/status_im/ui/screens/wallet/send/events.cljs (hunk 4 above)
Comment (dmitryn) [reply src/status_im/ui/screens/wallet/send/events.cljs]: This example is slightly different, but the idea that first line of threading macro doesn't start with operation, but with data structure / object

Comment (yenda) [reply src/status_im/ui/screens/wallet/send/events.cljs]: Ok I changed it in a way that should please both approaches.

http://tonsky.me/blog/readable-clojure/#dont-spare-names

**Comment by Serhy** [comment]
@yenda it was tested on build for Android: https://i.diawi.com/erouqL (f6b6d8e6+) and 
iOS: https://i.diawi.com/HcWTi8 (0.0.2162) and I can confirm it works fine. Transaction appears in history right away after it was signed in. 
I would like to cover the case when transaction from Unsigned list was Signed - and we able to do that now, since #3109 has merged recently.

Could you rebase this branch of develop, please?

**Comment by Serhy** [comment]
All good! Thanks a lot @yenda !

**Review by jeluard — APPROVED** [review]

**Review by goranjovic — APPROVED** [review]

**Review by dmitryn — APPROVED** [review]

**Pull request merged**